### PR TITLE
feat(terraform): combine tf plan and tf graphs for nested modules

### PR DESCRIPTION
**Pull request by achiar99**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

If nested modules and multi hops env is ON:
merge tf graph with tf plan graph to get correct graph checks for tf plan

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/terraform/plan_runner.py [hunk 1]
```diff
@@ -23,6 +23,7 @@
 from checkov.terraform.context_parsers.registry import parser_registry
 from checkov.terraform.plan_utils import create_definitions, build_definitions_context
 from checkov.terraform.runner import Runner as TerraformRunner, merge_reports
+from .deep_analysis_plan_graph_manager import DeepAnalysisGraphManager
```
Comment (gruebel) [inline checkov/terraform/plan_runner.py:26]:
```suggestion
from checkov.terraform.deep_analysis_plan_graph_manager import DeepAnalysisGraphManager
```

File: checkov/terraform/deep_analysis_plan_graph_manager.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        self.tf_graph: TerraformLocalGraph = tf_graph
+        self.tf_plan_graph: TerraformLocalGraph = tf_plan_graph
+        self._address_to_tf_vertex_map: Dict[str, TerraformBlock] = {}
+        self._address_to_tf_plan_vertex_map: Dict[str, TerraformBlock] = {}
+        self._apply_address_mapping()
+
+    def _apply_address_mapping(self) -> None:
+        for vertex in self.tf_graph.vertices:
+            if vertex.block_type == BlockType.RESOURCE:
+                self._address_to_tf_vertex_map[vertex.attributes[TF_PLAN_RESOURCE_ADDRESS]] = vertex
```
Comment (gruebel) [inline checkov/terraform/deep_analysis_plan_graph_manager.py:22]:
```suggestion
        self._address_to_tf_vertex_map = {
            vertex.attributes[TF_PLAN_RESOURCE_ADDRESS]: vertex
            for vertex in self.tf_graph.vertices
            if vertex.block_type == BlockType.RESOURCE:
        }
```
I think you can do it via a dict comprehension

File: checkov/terraform/deep_analysis_plan_graph_manager.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+        for vertex in self.tf_plan_graph.vertices:
```
Comment (gruebel) [inline checkov/terraform/deep_analysis_plan_graph_manager.py:24]: same here

File: checkov/terraform/deep_analysis_plan_graph_manager.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            if vertex.block_type == BlockType.RESOURCE:
+                self._address_to_tf_plan_vertex_map[vertex.attributes[TF_PLAN_RESOURCE_ADDRESS]] = vertex
+
+    def enrich_tf_graph_attributes(self) -> None:
+        for address, tf_plan_vertex in self._address_to_tf_plan_vertex_map.items():
+            tf_vertex = self._address_to_tf_vertex_map.get(address)
+            if not tf_vertex:
+                logging.warning(f'Cant find this address: {address} in tf graph')
```
Comment (gruebel) [inline checkov/terraform/deep_analysis_plan_graph_manager.py:32]:
```suggestion
                logging.info(f'Cant find this address: {address} in tf graph')
```
let's lower it to info or even debug for now

File: checkov/terraform/deep_analysis_plan_graph_manager.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                continue
+            tf_vertex.attributes = {**tf_vertex.attributes, **tf_plan_vertex.attributes}
+            tf_vertex.path = tf_plan_vertex.path
+
+    def filter_report(self, report: Report) -> None:
+        report.failed_checks = [check for check in report.failed_checks if
+                                check.resource_address in self._address_to_tf_plan_vertex_map]
+        report.passed_checks = [check for check in report.passed_checks if
+                                check.resource_address in self._address_to_tf_plan_vertex_map]
+        report.skipped_checks = [check for check in report.skipped_checks if
```
Comment (gruebel) [inline checkov/terraform/deep_analysis_plan_graph_manager.py:42]: can you also filter the `resources`, `extra_resources`, if we set there anything
also `parsing_errors` should only be the one from TF plan

**Review by gruebel — COMMENTED** [review]
nice, looks good so far

**Review by gruebel — APPROVED** [review]

**Comment by stromp** [comment]
@achiar99 @gruebel we encountered the issue that when we enable deep-analysis, the skipping of the checks (via code comment) is not working anymore. Is this intended behavior?

**Pull request merged**